avocado/utils/pci.py

```diff
@@ -0,0 +1,310 @@
+
+    :return: pci id of a pci device from sysfs.
+    """
+    path = "/sys/bus/pci/devices/%s" % device
+    if os.path.isdir(path):
+        path = "%s/%%s" % path
+        return ":".join(["%04x" % int(open(path % param).read(), 16)
+                         for param in ['vendor', 'device', 'subsystem_vendor',
+                                       'subsystem_device']])
+
+
+def get_slot_list():
+    """
+    Gets list of pci slots in the system.
+
+    :return: list of slots in the system.
+    """
+    return list(set(get_slot_from_sysfs(dev) for dev in get_pci_devices()))
+
+
+def get_specific_pci_id(device, part):
+    """
+    Gets specific pci id of given divice.
+
+    :parm device: pci device.
+    :parm part: part of pci id.
+
+    :return: specific pci id of a pci device.
+    """
+    cmd = "lspci -Dnvmm -s %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        for line in output.splitlines():
+            if part == line.split()[0].strip(':'):
+                return line.split()[-1]
+
+
+def get_pci_id(device):
+    """
+    Gets pci id of given divice.
+
+    :parm device: pci device.
+
+    :return: pci id of a pci device.
+    """
+    pci_id = []
+    for params in ['Vendor', 'Device', 'SVendor', 'SDevice']:
+        output = get_specific_pci_id(device, params)
+        if output:
+            pci_id.append(output)
+    if pci_id:
+        return ":".join(pci_id)
+
+
+def get_driver(device):
+    """
+    Gets the kernel driver in use of given device.
+
+    :parm device: pci device.
+
+    :return: driver of a pci device.
+    """
+    cmd = "lspci -ks %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        for line in output.splitlines():
+            if 'Kernel driver in use:' in line:
+                return line.split()[-1]
+
+
+def get_memory_address(device):
+    """
+    Gets the memory address of a pci device.
+    There may be multiple memory address for a pci device.
+    This function rerutns only the first such address.
+
+    :parm device: pci device.
+
+    :return: memory address of a pci device.
+    """
+    cmd = "lspci -bv -s %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        for line in output.splitlines():
+            if 'Memory at' in line:
+                return "0x%s" % line.split()[2]
+
+
+def get_mask(device):
+    """
+    Gets the mask of pci device.
+    There may be multiple memory entries for a pci device.
+    This mask is calculated only with the first such entry.
+
+    :parm device: pci device.
+
+    :return: mask of a pci device.
+    """
+    cmd = "lspci -vv -s %s" % device
+    output = process.system_output(cmd, shell=True, ignore_status=True)
+    if output:
+        dic = {'K': '10', 'M': '20', 'G': '30'}
+        for line in output.splitlines():
+            if 'Region' in line and 'Memory at' in line:
```

Comment:
Is the `Region` check necessary? I failed to found any `Memory at` without `Region` on my machine (I'm just wondering, no need to remove it as it might be needed on some archs).